### PR TITLE
portability: Towards C99/C11 compliance

### PR DESCRIPTION
**Pull request by mped-oticon**
This lessens the use of GNU extensions and permits code to compile with our Clang-based compiler.
Issues still remain, but these may require discussion; hopefully, the changes herein are acceptable. Code readability remains unharmed :)

Issue #9879.

### CODE REVIEW

File: CMakeLists.txt
Comment (SebastianBoe) [inline CMakeLists.txt]: -Werror=pointer-arith ?

Unless there is a good reason to allow pointer arithmetic during development ...

Comment (mped-oticon) [reply CMakeLists.txt]: Too aggressive? I did consider it but may lead to widespread breakage for some boards accustomed to GCC.

Comment (SebastianBoe) [reply CMakeLists.txt]: Warnings turn into errors in CI so the breakage should be the same.

File: subsys/bluetooth/host/hci_ecc.c
Comment (jhedberg) [inline subsys/bluetooth/host/hci_ecc.c]: The `buf->data` is already `u8_t *` so this particular cast looks redundant. The added parenthesis behind the pointer increment is still needed as a fix though (I think).

**Review by jhedberg — CHANGES_REQUESTED** [review]
Looks generally good. One small comment regarding a redundant typecast.

File: include/logging/log_backend.h
Comment (andyross) [inline include/logging/log_backend.h]: Likely to be a huge bike shed, but personally I'd prefer "char" for the type to use here, which is guaranteed by the standard to be the minimal memory unit size and readable to all.  Our oddball unit types are for unifying size (e.g. long vs. int) and signedness (e.g. char-is-unsigned-on-arm-wtf?).

But others feel probably even more strongly in the other direction, so just a note.

File: include/misc/sflist.h
Comment (andyross) [inline include/misc/sflist.h]: Removing all those semis from the top level is just asking to blow up our poor editors' indentation logic.  But I guess these are code-generation macros, so...

**Review by andyross — APPROVED** [review]
Not loving some of the aesthetics, but the issues here are unambiguous.  I still don't understand why a "clang-based" compiler is going to have trouble with this, unless it's a really really old version of clang.

File: include/logging/log_backend.h
Comment (ulfalizer) [reply include/logging/log_backend.h]: *Backseat driving*
What's with the `u8_t`, etc., types in Zephyr btw? Does Zephyr need to support pre-C99 compilers without `<stdint.h>`?

Comment (ulfalizer) [reply include/logging/log_backend.h]: It's a mish-mash in the Linux kernel too, but for something like Zephyr it might be reasonable to do a global search-replace I guess, unless there's reasons to avoid the `<stdint.h>` types.

Comment (andyross) [reply include/logging/log_backend.h]: We actually did use the standard types pervasively, and the existing types are the **result** of a global search and replace.  :)

I mostly stayed out of the fight, but yeah: inconsistent toolchain support for stdint.h was part of the reasoning.

File: include/misc/sflist.h
Comment (mped-oticon) [reply include/misc/sflist.h]: Macro hackery always deserves some punishment ;-)

File: include/logging/log_backend.h
Comment (mped-oticon) [reply include/logging/log_backend.h]: If char-is-unsigned-on-arm-wtf, then I would prefer u8_t cause I know it won't change signedness under my feet -- and it should, knock-on-wood, always be unsigned 8 bit for all supported platforms. Also for consistency.

**Comment by mped-oticon** [comment]
BTW, is there a way to invoke CI/shippable before spamming y'all?

File: CMakeLists.txt
Comment (pfalcon) [reply CMakeLists.txt]: Ideally, this should have been a separate commit. So, fixing true source-code errors was separated from debatable build system config change.

File: subsys/bluetooth/host/hci_ecc.c
Comment (pfalcon) [reply subsys/bluetooth/host/hci_ecc.c]: +1

**Comment by pfalcon** [comment]
@mped-oticon :

> BTW, is there a way to invoke CI/shippable before spamming y'all?

See previous comment.

**Comment by pfalcon** [comment]
(Maybe you meant something else. In this case, see scripts/{checkpatch.pl, sanitycheck}, etc.)

File: include/kernel.h
Comment (pfalcon) [inline include/kernel.h]: I'm not sure what happens here. You add a semicolon here to remove it in bunch of places and get the same result?

**Review by pfalcon — CHANGES_REQUESTED** [review]
Please confirm what happens here.

File: scripts/gen_syscalls.py
Comment (pfalcon) [inline scripts/gen_syscalls.py]: Please add to commit message explicit mention of what is done here. E.g.

> Also update scripts/gen_syscalls.py to not generate extra semicolons after K_SYSCALL_DECLARE* macros.


Again, if this was my project, I would request to split this commit in 2, both with focused commit messages, as it updates 2 widely different parts of the codebase.

**Review by pfalcon — CHANGES_REQUESTED** [review]
More detailed commit messages needed.

File: include/kernel.h
Comment (jukkar) [inline include/kernel.h]: It would probably be more clearer, if the commit message would explain why this change is needed. Now it just states what is being done but that is quite obvious by looking the code. The real issue is that if the macro is not defined, then there is extra semicolon inside the struct.

File: include/logging/log_backend.h
Comment (ceolin) [reply include/logging/log_backend.h]: I rather prefer using stdint, but if we use our types, u8_t is better than char, because it explicitly says that is is unsigned. Moreover, this is also a MISRA-C recommendation, if we want be compliant, we need to move in this direction.

File: include/kernel.h
Comment (ceolin) [reply include/kernel.h]: yeah, this commit makes no sense at all for me

**Comment by nashif** [comment]
regarding the ‘;’, I guess this is what you get when you build with -pedantic, I have seen this also with clang..

```
                                                                      ^
/home/nashif/github/zephyr/build/zephyr/include/generated/syscalls/kernel.h:69:87: warning: ISO C does not allow extra ‘;’ outside of a function [-Wpedantic]
 K_SYSCALL_DECLARE1_VOID(K_SYSCALL_K_QUEUE_INIT, k_queue_init, struct k_queue *, queue);
                                                                                       ^
/home/nashif/github/zephyr/build/zephyr/include/generated/syscalls/kernel.h:71:101: warning: ISO C does not allow extra ‘;’ outside of a function [-Wpedantic]
 K_SYSCALL_DECLARE1_VOID(K_SYSCALL_K_QUEUE_CANCEL_WAIT, k_queue_cancel_wait, struct k_queue *, queue);
                                                                                                     ^

```

File: include/kernel.h
Comment (nashif) [reply include/kernel.h]: isnt that in the commit message already?

`ISO C99 does not allow extra ‘;’ outside of a function`

Comment (jhedberg) [reply include/kernel.h]: Couldn't this confusing (and unusual) inclusion of a semicolon in the macro definition be avoided by having it expand to a dummy zero-length array element instead of nothing?

Comment (pfalcon) [reply include/kernel.h]: > isnt that in the commit message already?

> ISO C99 does not allow extra ‘;’ outside of a function

That commit message is possibly wrong (doesn't tell the whole truth).

Comment (pfalcon) [reply include/kernel.h]: > Couldn't this confusing (and unusual) inclusion of a semicolon in the macro definition

What's so confusing and unusual? In normal languages, macros are Turing-complete, imagine how that is "confusing". And in C, I didn't hear someone telling that the preprocessor as defined by some standard disallows "inclusion" of semicolon in macro body, references are welcome.

File: include/kernel.h
Comment (pfalcon) [reply include/kernel.h]: > It seems it's to avoid some warnings of dangling semicolons for the condition when the macro expands to nothing. 

Ah, d'oh. Too bad that we have 2 threads discussing the same line of code...

File: include/kernel.h
Comment (pfalcon) [reply include/kernel.h]: Ok, @jhedberg explains the issue here: https://github.com/zephyrproject-rtos/zephyr/pull/9917#discussion_r217603899 (too bad we have 2 branches of discussion of the same issue).

Which only shows that there can't be too detailed commit message, and majority of commit messages are terse up to confusion and doesn't described the changes they do well.

Comment (jhedberg) [reply include/kernel.h]: > I didn't hear someone telling that the preprocessor as defined by some standard disallows "inclusion" of semicolon in macro body, references are welcome.

@pfalcon I didn't say that it's disallowed. It's just that most macros in the code base don't do this, which to the casual reader makes it look like there's a bug in the code that's using the macro (since one expects to see a semicolon at the end of each struct member declaration). Anyway, I'm curious to hear thoughts on my zero-length array proposal that should help avoid all this.

Comment (pfalcon) [reply include/kernel.h]: > with which compiler issues fixed in this PR were discovered?

Ok, the description says "permits code to compile with our Clang-based compiler.". Is that public compiler? Based on recent Clang?

Anyway, I guess it's up to @mped-oticon to look at zero-sized array solution and test if it works/see if he'd be willing to use it.

(My concern with this PR overall remains not detailed enough commit messages.)

File: CMakeLists.txt
Comment (mped-oticon) [reply CMakeLists.txt]: @pfalcon Agree. Maybe even another, later, PR?

File: include/logging/log_backend.h
Comment (mped-oticon) [reply include/logging/log_backend.h]: Leaving as-is.

File: include/kernel.h
Comment (mped-oticon) [reply include/kernel.h]: @jhedberg zero-length array (aka flexible array member) is a fine idea/hack but has caveats: Look at `kernel.h`'s `struct k_stack` (or k_msgq or k_pipe), is the last member is some `u8_t flags`. That means our zero-length array is not the last member. This violates C99, see [1].

I would also be very skeptical of nesting structs that contain flexible array members - as the address for subsequent members would not be known at compile-time (which is required). I don't know if these struct nest though, but even if not, it could blow up later.

[1]
C99 spec: "As a special case, the **_last_** element of a structure with more than one named member may have an incomplete array type; this is called a flexible array member."

Comment (mped-oticon) [reply include/kernel.h]: @pfalcon 
> [...] looks confusing IMHO. So, doing reasonable steps to preserve them makes sense IMHO.
Hopefully the ALL_CAPS and parenthesis should inform you that this line is a macro, and therefore caution should be exercised. I agree it would be nice to keep up appearence, but IMHO the benefit (portability) outweighs the cons (readabillity). A popular RTOS targetting embedded systems will have ugly corners. Good thing we're not in a beauty contest :)

Comment (mped-oticon) [reply include/kernel.h]: @pfalcon I will add make the commit messages more detailed.

Comment (jhedberg) [reply include/kernel.h]: @mped-oticon I wasn't aware that this wasn't formally allowed elsewhere than at the end of structs. The main reason I thought it was is that there's a major project that uses it, namely the Linux kernel:

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/skbuff.h#n737
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/skbuff.h#n750
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/skbuff.h#n761

There are probably many more occurrences, but I pointed out these from the subsystem that I'm most familiar with. Anyway, because Linux uses zero-length arrays like this, it would seem bad business for a compiler not to support it, so I wonder how likely it is that we'd run into issues.

File: CMakeLists.txt
Comment (pfalcon) [reply CMakeLists.txt]: > Agree.

Thanks.

> Maybe even another, later, PR?

Whatever is easier for you. Just my 2 cents, I'd prefer it to be warning indeed, not error right away. As @SebastianBoe mentioned, warning will turn to errors during CI, so bad code won't leak into mainline by overlook. But during development/testing/debugging, sometimes it's helpful to take hacky shortcuts, so having it just as a warning is helpful.

File: include/kernel.h
Comment (pfalcon) [reply include/kernel.h]: >  I will make the commit messages more detailed

Thanks, appreciated!

Comment (mped-oticon) [reply include/kernel.h]: @pfalcon, @jhedberg It may work out with [0], but [] definitely does not. I thought zero-sized array and flexible arrays were treated the same, but I guess not :)

https://gcc.godbolt.org/z/1WEeEq 
removing `-pedantic` quenches the warning. So I guess [0] is fine if we never want `-pedantic` with `-Wall`, which would be a PITA.

That leaves me satisfied with the [0] proposal.

**Comment by andyross** [comment]
Just to dab a tiny bit more paint on the shed:  is this really just a pedantic-level warning from clang?  @mped-oticon, can you detail exactly what your compiler's misfeatures are?  If it's just an issue with warning flags vs. very-long-established-industry-convention-in-all-other-projects, maybe we don't need to care?

**Comment by nashif** [comment]
> Just to dab a tiny bit more paint on the shed: is this really just a pedantic-level warning from clang?

the above is from gcc.

The reason why we should care about standards, compilers behave differently, big difference between gcc and clang and the default behaviour, for one it is pedantic, for the other it is a warning that we would need to suppress, some compilers might just reduce to build the code, maybe not in this case, but I would try as much as possible be consistent and do not spend too much time thinking of how to solve it in zephyr build system working around the compilers.

The argument now is that it works as is so why bother, the counter argument is that someone will try some commercial compiler or some new version and things will break, we noticed that when we started build with llvm few years back.

**Comment by mped-oticon** [comment]
> @mped-oticon, can you detail exactly what your compiler's misfeatures are? 
Yes: We don't have support for GNU extensions in our Clang.

That means you will get this list incrementally, as each PR will contain what we need and not more.
We have already been exploring compiler options, permissive freestanding, etc, but without this PR, we get errors, not just warnings.

I'd be happy to elaborate more on our challenges with our current toolchain, but that would quickly become a long text :)

File: lib/libc/minimal/source/string/string.c
Comment (ceolin) [inline lib/libc/minimal/source/string/string.c]: ok, we may have a problem here, shouldn't you cast to `ssize_t` to avoid an underflow ? Instead of using `char *` it may be better using something `uint8_t` from `stdint.h` that is part of C standard and is recommended by other code guidelines.

File: lib/mempool/mempool.c
Comment (ceolin) [inline lib/mempool/mempool.c]: can't you use `uintptr_t` here ? It's more portable that `u32_t`

File: subsys/fs/nvs/nvs.c
Comment (ceolin) [inline subsys/fs/nvs/nvs.c]: shouldn't use `data8` here too ?

**Review by ceolin — COMMENTED** [review]
I'm 100% in favor of being compliant with C standard. It will make the code much more portable.

**Comment by superna9999** [comment]
ok for `drivers/flash/flash_stm32.c`

File: lib/libc/minimal/source/string/string.c
Comment (mped-oticon) [reply lib/libc/minimal/source/string/string.c]: > ok, we may have a problem here, shouldn't you cast to ssize_t to avoid an underflow ? 

Nope. I tried to make an ASCII drawing of why size_t is correct:
```
A):
    |            ssssssssssss  
    |      dddddddddddd         must be written forwards

                if ((size_t) (0 - 10) < 12) {
                if ((size_t) (-10) < 12) {
                if ((size_t) (-10) < 12) {
                if (4294967286 < 12) {      or some other big number
                if (false) {    False, so we do forward writes

B):
    |ssssssssssss              
    |      dddddddddddd         must be written backwards
 
                if ((size_t) (dest - src) < n) {
                if ((size_t) (10 - 0) < 12) {
                if ((size_t) 10 < 12) {
                if (10 < 12) {
                if (true) {     True, so we do backwards writes

C):
    |      dddddddddddd        dont care, could be no-op
    |      ssssssssssss
```

Comment (mped-oticon) [reply lib/libc/minimal/source/string/string.c]: > Instead of using char * it may be better using something uint8_t from stdint.h that is part of C standard and is recommended by other code guidelines.

I'm confortable with char*, but agree uint8_t* would be better. However I also know of  platforms having problems with stdint.h.

File: subsys/fs/nvs/nvs.c
Comment (mped-oticon) [reply subsys/fs/nvs/nvs.c]: Good catch. I'll fix to a consistent style.

File: lib/mempool/mempool.c
Comment (mped-oticon) [reply lib/mempool/mempool.c]: I don't see why. It has to become a pointer-to-u32_t regardless.
In this translation unit, u32_t is fine from a portability perspective as Zephyr promises to make that type available.

File: ext/hal/libmetal/libmetal/lib/io.c
Comment (pfalcon) [inline ext/hal/libmetal/libmetal/lib/io.c]: @mped-oticon, this makes changes to ext/hal/ , and that's usually a good reason to make that a separate commit, and review would usually include a question "did you submit these changes upstream"?

@galak: FYI

**Review by pfalcon — APPROVED** [review]
Thanks for all the refactors and detailed commit messages! +1 from me, but I found another concern: patching ext/hal/, which may need to be propagated upstream. Sorry if that causes additional delays, I'd personally like this merged asap!

**Comment by pfalcon** [comment]
@jhedberg : The cause of your -1 was fixed, can you re-review?

File: ext/hal/libmetal/libmetal/lib/io.c
Comment (mped-oticon) [reply ext/hal/libmetal/libmetal/lib/io.c]: @carlescufi It appears already fixed in libmetal: https://github.com/OpenAMP/libmetal/issues/65
Though, Zephyr's copy is not up to date.

File: ext/fs/nffs/src/nffs_write.c
Comment (mped-oticon) [reply ext/fs/nffs/src/nffs_write.c]: @carlescufi Thanks, I overlooked that one. Will do.

File: ext/hal/libmetal/libmetal/lib/io.c
Comment (mped-oticon) [reply ext/hal/libmetal/libmetal/lib/io.c]: @pfalcon See mine and Carles' comment above.
Should I update our copy of libmetal (where this has been fixed -- in a slightly different way than this PR)?

File: ext/fs/nffs/src/nffs_write.c
Comment (mped-oticon) [reply ext/fs/nffs/src/nffs_write.c]: @carlescufi PR created: https://github.com/apache/mynewt-nffs/pull/11
So I guess the process will be: Get mynewt-nffs PR merged, import fresh mynewt-nffs in Zephyr, rebase this PR. Correct?

**Comment by mped-oticon** [comment]
We need to reimport libmetal (should be OK) and mynewt-nffs (pending PR).
Should I do this in a separate Zephyr PR?

File: ext/hal/libmetal/libmetal/lib/io.c
Comment (pfalcon) [reply ext/hal/libmetal/libmetal/lib/io.c]: > Should I update our copy of libmetal (where this has been fixed -- in a slightly different way than this PR)?

Sure, that's the best solution, please coordinate with @galak who maintains Zephyr's copy. Also cc: @wjliang, upstream maintainer.

**Review by jhedberg — APPROVED** [review]

File: ext/fs/nffs/src/nffs_write.c
Comment (carlescufi) [reply ext/fs/nffs/src/nffs_write.c]: Correct. See the commit log in `ext/` to look at the format of commit messages that import external code.

**Comment by carlescufi** [comment]
@mped-oticon 

>Should I do this in a separate Zephyr PR?

It's fine if it's this PR, but separate commits.

File: ext/hal/libmetal/libmetal/lib/io.c
Comment (carlescufi) [reply ext/hal/libmetal/libmetal/lib/io.c]: Right, that's why we need @galak to import a new version. I believe he's already working towards that.

**Review by nashif — APPROVED** [review]

**Pull request merged**
